### PR TITLE
Display Finals Date From Default Term

### LINKED ISSUE #758: RFC: Finals date
It would be useful to see finals' dates in the finals tab in the calendar (next to the day), at least in the desktop version. I haven't looked at whether that's part of the API/WebSOC, but I think it should be pretty easy if it is. 

Maybe we can count back from the end dates of the quarters. Note: pay attention to how Summer session works.

Displaying the date and day with React Big Calendar *should* be as simple as passing flags in.

### PR DESCRIPTION
**Pull request by adcockdalton**
## Summary

1. Added new optional field to TermData.ts "finalsStartDate"
2. Changed calendarizeFinals to correctly assign finals to their respective dates instead of the arbitrary (2018, 0, 1)
3. Refactored CalendarRoot to use date instead of defaultDate
4. Dynamically update date

Regular classes still populate to the arbitrary 2018 date, which is simpler since those are displayed dateless. All other data relies on the finalsStartDate field specified by defaultTerm.

## Test Plan

1. Make a schedule with classes that have different finals, toggle back and forth.
2. Try with only Saturday finals classes (Math 2b) or a mix of each (any class and Math 2b), etc.

## Issues

Closes #758 

## Future Followup
Make an alternate to hardcoding future dates (#778) including for finals, or at least hardcode the next few years.

### CODE REVIEW

File: apps/antalmanac/src/stores/calendarizeHelpers.ts
Comment (KevinWu098) [inline apps/antalmanac/src/stores/calendarizeHelpers.ts:1]: Why is this file updated to use the `defaultFinalsStart()` values?

File: apps/antalmanac/src/lib/termData.ts
Comment (KevinWu098) [inline apps/antalmanac/src/lib/termData.ts:102]: If there's no `finalsStartDate`, I think it would be preferable to return `undefined`, rather that the hardcoded values

File: apps/antalmanac/src/components/Calendar/CalendarRoot.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/Calendar/CalendarRoot.tsx]:
```suggestion
    const date = defaultFinals ?? new Date(2018, 0, 1);
```
If finals aren't being shown, the value of date doesn't matter, so this can be simplified. 

_The code above is also written assuming you choose to return `undefined` from `getDefaultFinalsStartDate()` like described below. The unneeded `showFinalsSchedule` is still true no matter what, unless I've missed something (very possible!)._

File: apps/antalmanac/src/lib/termData.ts
Comment (KevinWu098) [inline apps/antalmanac/src/lib/termData.ts]:
```suggestion
function getDefaultFinalsStartDate(): Date | undefined {
    const [year, month, day] = termData[defaultTerm].finalsStartDate || [];

    return year && month && day ? new Date(year, month, day + 1) : undefined;
}
```
Additionally, it might be better to refactor so that you're not accessing a "magic" array -- all values should (ideally) tell you what they are

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Great PR, looking forward to using this feature.

Some minor refactoring suggestions and code questions, but otherwise looks great!

File: apps/antalmanac/src/components/Calendar/CalendarRoot.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/Calendar/CalendarRoot.tsx]:
```suggestion

    const calendarStyling = isMobile ? { height: `calc(100% - 55px)` } : { height: `calc(100vh - 104px)` };
    
    const calendarTimeFormat = isMilitaryTime ? 'HH:mm' : 'h:mm A';
    const calendarGutterTimeFormat = isMilitaryTime ? 'HH:mm' : 'h A';
    
    const defaultFinals = getDefaultFinalsStartDate();
    const finalsDateFormat = defaultFinals ? 'ddd MM/DD' : 'ddd';
    const date = showFinalsSchedule ? defaultFinals : new Date(2018, 0, 1);
```

Honestly, this is a total nitpick (not your fault/not related to this PR), and could be a code smell, but I'd like to break up some of these values so it's more immediately clear what is related to what.

File: apps/antalmanac/src/stores/calendarizeHelpers.ts
Comment (adcockdalton) [reply apps/antalmanac/src/stores/calendarizeHelpers.ts:1]: Apologies if I was not clear enough on point 2; I can explain more on that: 

RBC, as we are implementing it now, populates the calendar with events that are assigned specific days. Prior to this PR, every event, class, final etc. populates on the first week of January 2018 for whatever reason. However, to get the finals to display with the correct actual dates, we have to set the date (by date, I mean the value assigned to the date variable passed into the Calendar object) to the start date of finals (adjusted by 1 to show the correct week regardless of a Saturday final) so that the current viewed week is the correct finals week when toggled rather than the first week of January 2018. That's just how RBC works unfortunately... I added a couple comments to the code for posterity.

So what I did is, if the defaultFinalsStart exists, is assign the finals to the actual week they occur on. Then if showFinalsSchedule is on you will actually see the dates.

This is also why I can't simplify the logic with showFinalsSchedule like you suggested in the other comment: if we do so, then you will not see any classes because the classes only show up in 2018.

lmk if there are any more questions on that, it was annoying figuring it all out haha

Comment (adcockdalton) [reply apps/antalmanac/src/stores/calendarizeHelpers.ts:1]: making a push with the other changes you suggested in about 20 minutes once I finish a quiz 👍

Comment (adcockdalton) [reply apps/antalmanac/src/stores/calendarizeHelpers.ts:1]: If it helps to demo, observe the default behavior by removing the hardcoded finals date from the winter 2024 term in termdata.ts

Comment (KevinWu098) [reply apps/antalmanac/src/stores/calendarizeHelpers.ts:1]: Gotcha, thanks for explaining it to me

**Review by KevinWu098 — APPROVED** [review]
LGTM, thanks for working on this! 🚀

There's probably some more discussion to be had about how to handle other terms than the current one, but that's out of scope for this PR (and is likely gated behind Google Auth).

I'll approve but not merge for now, as I want to confirm the final implementation at the sprint this Wednesday

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Whoops! Sike, not LGTM.

@adcockdalton, you get to now enjoy writing tests. The `calendarize-helpers` test is failing (should be quick fix), but that'll need to get done before we can merge.

**Review by KevinWu098 — APPROVED** [review]
LGTM; bit loathe to use a hardcoded test for a changing value but I can live with it (we'll just deal with it later TM)

**Pull request merged**
